peel_payment_onion static fn in channelmanager

Evanfeenstra opened the pull request:
Adds a public `peel_payment_onion` function that takes `msgs::UpdateAddHTLC` and returns `PendingHTLCInfo`.

I separated the internals of `decode_update_add_htlc_onion`, `construct_recv_pending_htlc_info`, and `construct_fwd_pending_htlc_info` into static functions, and used them to make the `PendingHTLCInfo`. There are two validation closures passed to `peel_payment_onion`... some of the outbound channel checking is done in these closures, so its up to the caller to validate against current channel state.

I am unsure about `phantom_shared_secret` and `allow_underpay` in `create_recv_pending_htlc_info` so i left them None/false when called from `peel_payment_onion`

Not sure if this is the best approach! But its much better than https://github.com/lightningdevkit/rust-lightning/pull/2677

Evanfeenstra commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2903,374 +2920,106 @@ where
+			}
+		};
+
+		let validate_peer_chan = |
```
Ok, its much cleaner now. I repeated the CLTV checks in the `peel_payment_onion` fn though, but I guess that's fine.

Also added a test for `peel_payment_onion`

Evanfeenstra commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7735,320 @@ where
+				if msg.cltv_expiry > cur_height + CLTV_FAR_FAR_AWAY as u32 { // expiry_too_far
+					return Err("CLTV expiry is too far in the future");
+				}
+				if (outgoing_cltv_value) as u64 <= (cur_height + LATENCY_GRACE_PERIOD_BLOCKS) as u64 {
```
ok! got it figured out so there is no redundancy

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2903,192 +2920,28 @@ where
-		})
-	}
-
 	fn construct_recv_pending_htlc_info(
```
Can't we just get rid of this method now, since the only thing it does is call a different method?

Evanfeenstra replied:
ok ya, i just got rid of it and call `create_recv_pending_htlc_info` from the tests too

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3103,37 +2956,13 @@ where
-			}
-		};
+		// it is a receive, so no need for outbound checks
+		if let None = next_packet_details_opt {
```
nit: this is more idiomatic, same applies below 
```suggestion
		if next_packet_details_opt.is_none() {
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3103,37 +2956,13 @@ where
+		}
+		let (next_packet_pk, outgoing_scid, outgoing_amt_msat, outgoing_cltv_value) = 
+			next_packet_details_opt.unwrap();
+		let next_packet_pk_opt = Some(next_packet_pk);
```
nit: let's get rid of this variable and just add the `Some(..)` wrapping on return

Evanfeenstra replied:
gotcha

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12044,6 +12194,147 @@ mod tests {
+		};
+	}
+
+	fn make_udpate_add_msg(
```
```suggestion
	fn make_update_add_msg(
```

Evanfeenstra replied:
oops lol

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7714,348 @@ where
+}
+
+fn check_incoming_htlc_cltv(
+	cur_height: u32, outgoing_cltv_value: u32, cltv_expiry: u32, check_min_cltv_expiry_delta: bool
```
Let's get rid of `check_min_cltv_expiry_delta` for now and just always check to reduce complexity

Evanfeenstra replied:
sounds good, done

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7714,348 @@ where
+	Ok((next_hop, shared_secret, Some(next_packet_details)))
+}
+
+fn check_incoming_htlc_cltv(
```
In the future, a change like abstracting this logic into a method should go in its own commit. This commit is kinda hard to review since it does a lot of code moves at once.

Evanfeenstra replied:
noted

valentinewallace reviewed (commented):
Still making my way through

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+				msg.onion_routing_packet.public_key.unwrap(), &shared_secret);
+			(next_packet_pk, short_channel_id, amt_to_forward, outgoing_cltv_value)
+		},
+		// We'll do receive checks in [`Self::construct_pending_htlc_info`] so we have access to the
+		// inbound channel's state.
```
This comment doesn't really make sense anymore, can be removed in a separate commit.

Evanfeenstra replied:
ok done

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+/// acceptance. If the payment is to be received, and the amount matches the expected amount for
+/// a given invoice, this indicates the UpdateAddHTLC, once fully committed in the channel, will
+/// generate an Event::PaymentReceived.
+pub fn peel_payment_onion<
+	NS: Deref,
+	L: Deref,
+	T: secp256k1::Verification,
+>(
```
nit: this can be one-line, similar with `decode_incoming_update_add`

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+	})
+}
+
+/// Peel one layer off an incoming onion, returning PendingHTLCInfo (either Forward or Receive).
+/// This does all the relevant context-free checks that LDK requires for payment relay or
+/// acceptance. If the payment is to be received, and the amount matches the expected amount for
+/// a given invoice, this indicates the UpdateAddHTLC, once fully committed in the channel, will
+/// generate an Event::PaymentReceived.
```
Use ticks around code identifiers like `PendingHTLCInfo` and others here, and also add doc links for `PendingHTLCInfo` and the event.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+		decode_incoming_update_add_htlc_onion(msg, node_signer, logger, secp_ctx
+	).map_err(|_| InboundOnionErr {
+		msg: "Failed to decode update add htlc onion",
+		err_code: 0x4000 | 22,
+		err_data: Vec::new(),
```
Rather than hardcoding the error code/data here, we should extract those fields out of the error returned from `decode_incoming_update_add`.

Evanfeenstra replied:
This one is tough... since HTLCFailureMsg::Relay code and data is encrypted. But i went ahead and extracted what i could easily

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+		err_data: Vec::new(),
+	})?;
+	Ok(match hop {
+		onion_utils::Hop::Forward{ next_hop_data, next_hop_hmac, new_packet_bytes } => {
```
nit:
```suggestion
		onion_utils::Hop::Forward { next_hop_data, next_hop_hmac, new_packet_bytes } => {
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+	NS::Target: NodeSigner,
+	L::Target: Logger,
+{
+	let (hop, shared_secret, next_packet_details_opt) = 
```
nit: trailing whitespace

Evanfeenstra replied:
ugh i'll be more careful with whitespace

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+) -> Result<(
+	onion_utils::Hop, [u8; 32], Option<(Result<PublicKey, secp256k1::Error>, u64, u64, u32)>
+), HTLCFailureMsg> 
```
nit: trailing whitespace

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+) -> Result<(
+	onion_utils::Hop, [u8; 32], Option<(Result<PublicKey, secp256k1::Error>, u64, u64, u32)>
```
`next_packet_details` should really be made into a struct rather than a long optional tuple. Can be saved for follow-up though.

Evanfeenstra replied:
I went ahead and did it! Simple change

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,346 @@ where
+	})?;
+	Ok(match hop {
+		onion_utils::Hop::Forward{ next_hop_data, next_hop_hmac, new_packet_bytes } => {
+			if next_packet_details_opt.is_none() {
+				// Forward should always include the next hop details
+				return Err(InboundOnionErr {
+					msg: "Failed to decode update add htlc onion",
+					err_code: 0x4000 | 22,
+					err_data: Vec::new(),
+				});
+			};
+			let (next_packet_pubkey, _, _, outgoing_cltv_value) = next_packet_details_opt.unwrap();
```
I prefer doing it this way to avoid an `unwrap`, even though it's safe (similar elsewhere):
```suggestion
			let (next_packet_pubkey, outgoing_cltv_value) = match next_packet_details_opt {
				Some((pk, _, _, cltv)) => (pk, cltv),
				None => {
					// Forward should always include the next hop details
					return Err(InboundOnionErr {
						msg: "Failed to decode update add htlc onion",
						err_code: 0x4000 | 22,
						err_data: Vec::new(),
					});
				}
			};
```

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3103,37 +2943,12 @@ where
-			}
-		};
+		// it is a receive, so no need for outbound checks
+		if next_packet_details_opt.is_none() {
+			return Ok((next_hop, shared_secret, None));
+		}
+		let (next_packet_pk, outgoing_scid, outgoing_amt_msat, outgoing_cltv_value) = 
+			next_packet_details_opt.unwrap();
```
Similar here regarding avoiding an `unwrap` with `match`

Evanfeenstra commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12044,6 +12188,147 @@ mod tests {
+		let preimage_bytes = [43; 32];
+		let preimage = PaymentPreimage(preimage_bytes);
+		let rhash = Sha256::hash(&preimage_bytes).to_vec();
+		let rhash_bytes: [u8; 32] = rhash.try_into().unwrap();
```
hm ya thats weird, thanks for catching it

Evanfeenstra commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3207,40 +3022,19 @@ where
 				if let Err((err, code)) = chan.htlc_satisfies_config(&msg, outgoing_amt_msat, outgoing_cltv_value) {
 					break Some((err, code, chan_update_opt));
 				}
+				// min_cltv_expiry_delta is already checked above in htlc_satisfies_config
```
oh ya, that comment doesn't make sense any more

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,351 @@ where
+
+fn decode_incoming_update_add_htlc_onion<NS: Deref, L: Deref, T: secp256k1::Verification>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+) -> Result<(
+	onion_utils::Hop, [u8; 32], Option<NextPacketDetails>
+), HTLCFailureMsg>
```
Can be one line now

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12044,6 +12194,137 @@ mod tests {
 	}
+
+	#[test]
+	fn test_peel_payment_onion() {
```
Not to harp on this but good amount of trailing whitespace in these tests/utils

Evanfeenstra replied:
ok, pretty i got em all

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,351 @@ where
+	cur_height: u32, outgoing_cltv_value: u32, cltv_expiry: u32
+) -> Result<(), (&'static str, u16, bool)> {
+	if (cltv_expiry as u64) < (outgoing_cltv_value) as u64 + MIN_CLTV_EXPIRY_DELTA as u64 {
+		// We really should set `incorrect_cltv_expiry` here but as we're not
+		// forwarding over a real channel we can't generate a channel_update
+		// for it. Instead we just return a generic temporary_node_failure.
```
Rather than keeping this behavior, we should always return the correct error in this method and convert it if needed at the callsite:

```diff
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index e9433a199..c4a37c80a 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -3030,10 +3030,16 @@ where

                        let cur_height = self.best_block.read().unwrap().height() + 1;

-                       if let Err((err_msg, code, include_chan_update_opt)) = check_incoming_htlc_cltv(
+                       if let Err((err_msg, code)) = check_incoming_htlc_cltv(
                                cur_height, outgoing_cltv_value, msg.cltv_expiry
                        ) {
-                               let chan_update_opt = if include_chan_update_opt { chan_update_opt } else { None };
+                               if code & 0x1000 != 0 && chan_update_opt.is_none() {
+                                       // We really should set `incorrect_cltv_expiry` here but as we're not
+                                       // forwarding over a real channel we can't generate a channel_update
+                                       // for it. Instead we just return a generic temporary_node_failure.
+                                       break Some((err_msg, 0x2000 | 2, None))
+                               }
+                               let chan_update_opt = if code & 0x1000 != 0 { chan_update_opt } else { None };
                                break Some((err_msg, code, chan_update_opt));
                        }

@@ -8015,14 +8021,11 @@ where

 fn check_incoming_htlc_cltv(
        cur_height: u32, outgoing_cltv_value: u32, cltv_expiry: u32
-) -> Result<(), (&'static str, u16, bool)> {
+) -> Result<(), (&'static str, u16)> {
        if (cltv_expiry as u64) < (outgoing_cltv_value) as u64 + MIN_CLTV_EXPIRY_DELTA as u64 {
-               // We really should set `incorrect_cltv_expiry` here but as we're not
-               // forwarding over a real channel we can't generate a channel_update
-               // for it. Instead we just return a generic temporary_node_failure.
                return Err((
                                "Forwarding node has tampered with the intended HTLC values or origin node has an obsolete cltv_expiry_delta",
-                               0x2000 | 2, false,
+                               0x1000 | 13, // incorrect_cltv_expiry
                ));
        }
        // Theoretically, channel counterparty shouldn't send us a HTLC expiring now,
```

Note that we also don't need to return whether a channel update is needed, we can detect it based on whether the error code includes `0x1000`.

Evanfeenstra replied:
cool

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7906,6 +7704,351 @@ where
+pub fn peel_payment_onion<NS: Deref, L: Deref, T: secp256k1::Verification>(
+	msg: &msgs::UpdateAddHTLC, node_signer: &NS, logger: &L, secp_ctx: &Secp256k1<T>,
+	cur_height: u32, accept_mpp_keysend: bool,
+) -> Result<PendingHTLCInfo, InboundOnionErr>
```
The error we return at the moment is a bit hard to use for failing -- callers have to parse it to figure out whether they should be failing back with `update_fail` or `update_fail_malformed`. 

In separate commit(s), could we add a new error enum `PeelOnionError` with `Fail` and `FailMalformed` variants that each contain an error code + data, and have `decode_incoming_update_add_htlc_onion` + `peel_onion` return this new error per your offline suggestion?

Evanfeenstra replied:
ok, the `Fail` variant also includes the shared_secret so the failure reason can be encrypted by the caller

valentinewallace reviewed (commented):
Almost there!

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -106,47 +106,63 @@ use crate::ln::script::ShutdownScript;
+	pub routing: PendingHTLCRouting,
+	/// Shared secret from the previous hop.
+	pub incoming_shared_secret: [u8; 32],
 	payment_hash: PaymentHash,
```
Why leave this as the only unexposed field?

valentinewallace reviewed (commented):
Needs rebase unfortunately :( 

LGTM after this. Also fine with fixing these in follow-up.

Evanfeenstra commented:
> Needs rebase unfortunately :(
> 
> LGTM after this. Also fine with fixing these in follow-up.

OK, happy to do a follow-up on these items as soon as it gets merged

Evanfeenstra commented:
> Rather than using merge commits to rebase, please use rebase directly. See https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes.

Oh, thanks for the tip. I reverted the merge commit. Also reverted the PeelOnionError stuff for a follow-up PR instead, per @valentinewallace's suggestion

valentinewallace commented:
Having done more research into CLN and LND's htlc interception APIs, I don't think we need to follow-up  adding `PeelOnionError`. 

Both APIs take either an error code *or* a decoded onion error packet, so our current error should be sufficient for that. 

I was also thinking the LSPS4 spec may need better error info, but from talking to @TheBlueMatt it sounds like the LSP would always just error with "unknown next channel." 

So excuse the misleading direction, I think this PR is fine as-is.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -378,7 +394,8 @@ impl HTLCSource {
-struct InboundOnionErr {
+/// Invalid inbound onion payment.
+pub struct InboundOnionErr {
 	err_code: u16,
 	err_data: Vec<u8>,
 	msg: &'static str,
```
Oops, these fields should be made `pub`.

Evanfeenstra replied:
ok, done
